### PR TITLE
[ADD] stock_move_planned_consumed_date

### PR DESCRIPTION
**Pull request by petrus-v**
Base module to avoid mixing move line that is planned to consumed in different dates

### CODE REVIEW

**Comment by rousseldenis** [comment]
@petrus-v Have you tried with date_deadline field ?

**Comment by rousseldenis** [comment]
> @petrus-v Have you tried with date_deadline field ?

Which is propagated at sale order confirmation from the commitment_date on the order level

**Comment by petrus-v** [comment]
> @petrus-v Have you tried with date_deadline field ?

I know about the date_deadline which propagate on stock_picking, but if you check the availability of those picking a bit earlier nothing prevent to reserved products that will be expired before the deadline (At the moment we haven't installed stock release modules mechanisms), also in complex industry with multiple warehouse where procurement group are not necessarily propagated we could get some transfer (stock.picking) with different moves where expected consumed date are different (to make this easier to reproduce we are using https://github.com/oca/stock-logistics-workflow/pull/1226 which basically round the deadline date time to date)

this is the base module for #1509 and #1511

**Comment by rousseldenis** [comment]
> > @petrus-v Have you tried with date_deadline field ?
> 
> I know about the date_deadline which propagate on stock_picking, but if you check the availability of those picking a bit earlier nothing prevent to reserved products that will be expired before the deadline (At the moment we haven't installed stock release modules mechanisms), also in complex industry with multiple warehouse where procurement group are not necessarily propagated we could get some transfer (stock.picking) with different moves where expected consumed date are different (to make this easier to reproduce we are using #1226 which basically round the deadline date time to date)
> 
> this is the base module for #1509 and #1511

Ok, understand.

FYI, there is a change in v16 (and v15) merged yesterday that allows to ignore date_deadline when merging: https://github.com/odoo/odoo/commit/fdaba15c4f83f75fa2bc16c8e44d9c7093239ee3

File: stock_move_planned_consumed_date/readme/CONTRIBUTIRS.rst
Comment (alexandregaldeano) [inline stock_move_planned_consumed_date/readme/CONTRIBUTIRS.rst:1]: `CONTRIBUTIRS`? Are you sure? 😛

File: stock_move_planned_consumed_date/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        "Planned consumed date",
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+        help="This is the expected consumed/usage date of the product by the customer.",
+    )
+
+    def _prepare_procurement_values(self):
+        vals = super()._prepare_procurement_values()
+        vals["planned_consumed_date"] = self.planned_consumed_date
+        return vals
```
Comment (alexandregaldeano) [inline stock_move_planned_consumed_date/models/stock_move.py]:
```suggestion
        return {
            **super()._prepare_procurement_values(),
            "planned_consumed_date": self.planned_consumed_date,
        }
```

WDYT?

File: stock_move_planned_consumed_date/models/stock_move.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    @api.model
+    def _prepare_merge_moves_distinct_fields(self):
+        distinct_fields = super()._prepare_merge_moves_distinct_fields()
+        distinct_fields.append("planned_consumed_date")
+        return distinct_fields
```
Comment (alexandregaldeano) [inline stock_move_planned_consumed_date/models/stock_move.py]:
```suggestion
        return [*super()._prepare_merge_moves_distinct_fields(), "planned_consumed_date"]
```

WDYT?

File: stock_move_planned_consumed_date/models/stock_rule.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+from odoo import models
+
+
+class StockRule(models.Model):
+    _inherit = "stock.rule"
+
+    def _get_custom_move_fields(self):
+        fields = super()._get_custom_move_fields()
+        fields += ["planned_consumed_date"]
+        return fields
```
Comment (alexandregaldeano) [inline stock_move_planned_consumed_date/models/stock_rule.py]: Same here

**Review by alexandregaldeano — APPROVED** [review]
My comments are about code style preference, as such, you can of course discard them :)

File: stock_move_planned_consumed_date/models/stock_move.py (hunk 1 above)
Comment (petrus-v) [reply stock_move_planned_consumed_date/models/stock_move.py]: done

File: stock_move_planned_consumed_date/models/stock_move.py (hunk 2 above)
Comment (petrus-v) [reply stock_move_planned_consumed_date/models/stock_move.py]: done

File: stock_move_planned_consumed_date/models/stock_rule.py (hunk 3 above)
Comment (petrus-v) [reply stock_move_planned_consumed_date/models/stock_rule.py]: yeah !! thanks

File: stock_move_planned_consumed_date/views/stock_picking_views.xml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+                expr="//page[@name='operations']/field[@name='move_ids_without_package']/tree/field[@name='reserved_availability']"
+                position="after"
+            >
+                <field name="planned_consumed_date" optional="hide" />
+            </xpath>
+            <xpath
+                expr="//page[@name='operations']/field[@name='move_ids_without_package']/form//field[@name='reserved_availability']"
+                position="after"
+            >
+                <field name="planned_consumed_date" optional="hide" />
```
Comment (petrus-v) [inline stock_move_planned_consumed_date/views/stock_picking_views.xml]:
```suggestion
                <field name="planned_consumed_date" />
```

Comment (rousseldenis) [reply stock_move_planned_consumed_date/views/stock_picking_views.xml]: @petrus-v I would say it should be optional="show" instead as tree view can be very overcharged.

**Pull request closed without merging**